collector/topological_inventory.py

```diff
@@ -0,0 +1,255 @@
+    fk_name (str)
+        Column where the Foreign Key is located
+    fk_id (str)
+        Foreign Key value
+
+    Returns
+    -------
+    list
+        Updated page_data
+
+    """
+    if not (fk_name and fk_id):
+        return page_data
+
+    for row in page_data:
+        row[fk_name] = fk_id
+    return page_data
+
+
+def _collect_data(url: str, fk_name: str = None,
+                  fk_id: str = None, headers: dict = None) -> dict:
+    """Aggregate data from all pages.
+
+    Returns data aggregated from all pages together
+
+    Parameters
+    ----------
+    url (str)
+        URI to the first page, where to start the traverse
+    fk_name (str)
+        Foreign key column to update if needed
+    fk_id (str)
+        Foreign key value to update if needed
+
+    Returns
+    -------
+    dict
+        All data aggregated across all pages
+
+    Raises
+    ------
+    utils.RetryFailedError
+        Connection failed, data is not complete
+
+    """
+    # Collect data from the first page
+    prometheus_metrics.METRICS['gets'].inc()
+    resp = utils.retryable('get', url, headers=headers)
+    prometheus_metrics.METRICS['get_successes'].inc()
+    resp = resp.json()
+    all_data = resp['data']
+
+    # Walk all pages
+    while resp['links'].get('next'):
+        prometheus_metrics.METRICS['gets'].inc()
+        resp = utils.retryable(
+            'get',
+            f'{TOPOLOGICAL_INVENTORY_HOST}{resp["links"]["next"]}',
+            headers=headers
+        )
+        resp = resp.json()
+        prometheus_metrics.METRICS['get_successes'].inc()
+        all_data += resp['data']
+
+    return _update_fk(all_data, fk_name, fk_id)
+
+
+def _query_main_collection(collection: str, headers: dict = None) -> dict:
+    """Query a Collection.
+
+    Parameters
+    ----------
+    collection (str)
+        A collection to download
+
+    Returns
+    -------
+    dict
+        All data aggregated across all pages
+
+    Raises
+    ------
+    utils.RetryFailedError
+        Connection failed, data is not complete
+
+    """
+    return _collect_data(f'{BASE_URL}/{collection}', headers=headers)
+
+
+def _query_sub_collection(
+        main_collection: str,
+        sub_collection: str,
+        data: dict,
+        foreign_key: str,
+        headers: dict = None
+) -> dict:
+    """Query a SubCollection for all records in the main collection.
+
+    Parameters
+    ----------
+    main_collection (str)
+        A parent collection pointer
+    sub_collection (str)
+        A collection to download
+    data (dict)
+        Already available data for reference
+    foreign_key (str)
+        Foreign key column to be populated in sub-collection records
+
+    Returns
+    -------
+    dict
+        All data aggregated across all pages
+
+    Raises
+    ------
+    utils.RetryFailedError
+        Connection failed, data is not complete
+
+    """
+    url = f'{BASE_URL}/{main_collection}/{{}}/{sub_collection}'
```

Comment:
@tumido So the issue is here on this line and 170

if `main_collection` is something like `container_groups?filter[archived_at][nil]`, then the url won't be correct.

We need `main_collection` to be a single entity word like  `container_groups`

Locally here, just for testing purposes, I have resolved it with -

```py
main_collection_minus_filter = main_collection.replace('?filter[archived_at][nil]', '')

url = f'{BASE_URL}/{main_collection_minus_filter}/{{}}/{sub_collection}'
```

and 

```py
for item in data[main_collection_minus_filter]:
```

How to make this more clean, I leave that to you.
Maybe we should not be combining entity and filter in `main_collection`, and that they should be separate, just so that we don't have to do any string manipulation.

Reply:
I see, I'll fix that, thanks! :slightly_smiling_face:

Reply:
@AparnaKarve the fix is pretty easy, in the `topological_queries.yml` I had the `containers` subcollection defined wrongly. Fixed that.

Reply:
Good to know that was easy... sorry I completely missed looking at `topological_queries.yml` and yes ...it's looking good!! :)